doc/TAM/SAI-Proposal-TAM-stream-telemetry.md

```diff
@@ -0,0 +1,798 @@
+    - [Netlink Module](#netlink-module)
+  - [Data format](#data-format)
+    - [IPFIX header](#ipfix-header)
+    - [IPFIX template](#ipfix-template)
+    - [IPFIX data](#ipfix-data)
+    - [Netlink message](#netlink-message)
+  - [Bandwidth Estimation](#bandwidth-estimation)
+  - [SAI API](#sai-api)
+    - [Creating HOSTIF object](#creating-hostif-object)
+    - [Creating HOSTIF trap group](#creating-hostif-trap-group)
+    - [Creating HOSTIF user defined trap](#creating-hostif-user-defined-trap)
+    - [Creating Hostif table entry](#creating-hostif-table-entry)
+    - [Creating TAM transport object](#creating-tam-transport-object)
+    - [Creating TAM collector object](#creating-tam-collector-object)
+    - [Creating TAM report object](#creating-tam-report-object)
+    - [Creating TAM telemetry type object](#creating-tam-telemetry-type-object)
+    - [Creating TAM telemetry object](#creating-tam-telemetry-object)
+    - [Creating TAM counter subscription objects](#creating-tam-counter-subscription-objects)
+    - [Creating TAM object](#creating-tam-object)
+    - [Query IPFIX template](#query-ipfix-template)
+    - [Enable/Disable telemetry stream](#enabledisable-telemetry-stream)
+
+## Revision
+
+| Rev | Date       | Author | Change Description |
+| --- | ---------- | ------ | ------------------ |
+| 0.1 | 09/06/2024 | Ze Gan | Initial version    |
+
+## Scope
+
+This document outlines the high-level design of stream telemetry, focusing primarily on the internal aspects of device rather than external telemetry systems.
+
+## Definitions/Abbreviations
+
+| Abbreviation | Description                               |
+| ------------ | ----------------------------------------- |
+| SAI          | The Switch Abstraction Interface          |
+| IPFIX        | Internet Protocol Flow Information Export |
+| TAM          | Telemetry and Monitoring                  |
+| BW           | Bandwidth                                 |
+
+## Overview
+
+The existing telemetry solution relies on a process to proactively query stats and counters via the SAI API. This approach causes the process to spend excessive time on SAI communication. The stream telemetry described in this document aims to provide a more efficient method for collecting object stats. The main idea is that selected stats will be proactively pushed from the vendor's driver to the collector via netlink.
+
+## Requirements
+
+- The number of SAI object types should not exceed 32,768 ($2^{15}$). This means the value of SAI_OBJECT_TYPE_MAX should be less than 32,768.
+- The number of SAI object extension types should not exceed 32,768.
+- The number of stats types for a single SAI object type should not exceed 32,768.
+- The number of extension stats types for a single SAI object type should not exceed 32,768.
+- The number of SAI objects of the same type should not exceed 32,768.
+- The vendor SDK should support publishing stats in IPFIX format and its IPFIX template.

```

Comment:
Because the IPFIX recording is published by the vendor, So, according to the self-consistency principle, the IPFIX template will also be published from the vendor side as well. For example, we don't limit the order of data in the IPFIX , it's acceptable once the order of date is consistent in the IPFIX template or recording.
We expect the generation of all IPFIX template or reocrding is maintained by one side.